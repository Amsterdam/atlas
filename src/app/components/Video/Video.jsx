/* eslint-disable jsx-a11y/media-has-caption */

import React from 'react';
import PropTypes from 'prop-types';
import './Video.scss';

class Video extends React.Component {
  componentDidUpdate(prevProps) {
    const { videoPlayer, props } = this;
    const { play, position } = props;

    // Determine to play or pause the player
    if (prevProps.play !== play) {
      if (play) {
        videoPlayer.play();
      } else {
        videoPlayer.pause();
      }
    }

    // Player position
    if (videoPlayer.currentTime !== position) {
      videoPlayer.currentTime = position;
    }
  }

  render() {
    const { src, poster, type } = this.props;
    return (
      <video
        ref={(c) => {
          this.videoPlayer = c;
        }}
        data-object-fit
        preload="metadata"
        muted="true"
        className="c-video__element"
        poster={poster}
        loop
      >
        <source
          {...{
            src,
            type
          }}
        />
      </video>
    );
  }
}

Video.defaultProps = {
  play: false,
  position: 0
};

Video.propTypes = {
  src: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  poster: PropTypes.string.isRequired,
  play: PropTypes.bool,
  position: PropTypes.number
};

export default Video;
