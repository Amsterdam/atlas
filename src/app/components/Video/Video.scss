@import '../../../shared/styles/config';

.c-video {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c-video__element {
  height: 100%;
  object-fit: cover;
  width: 100%;

  video {
    // turn off play controls in iOS: needs vendor prefixes
    // sass-lint:disable no-vendor-prefixes
    // sass-lint:selector-pseudo-element-no-unknown
    ::-webkit-media-controls-panel {
      -webkit-appearance: none;
      display: none !important;
    }

    ::-webkit-media-controls-play-button {
      -webkit-appearance: none;
      display: none !important;
    }

    ::-webkit-media-controls-start-playback-button {
      -webkit-appearance: none;
      display: none !important;
    }
  }
  // sass-lint:enable no-vendor-prefixes
}
