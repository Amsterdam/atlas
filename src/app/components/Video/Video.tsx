/* eslint-disable jsx-a11y/media-has-caption */

import { Component, FunctionComponent, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'

interface VideoProps {
  src?: string
  type?: string
  poster?: string
  play?: boolean
  showControls?: boolean
  position?: number
}

const Vidiho: FunctionComponent<VideoProps> = ({
  src,
  poster,
  type,
  showControls,
  children,
  play,
  position,
}) => {
  const videoPlayer = useRef<HTMLVideoElement>(null)

  // useEffect(() => {
  //   // Determine to play or pause the player
  //   if (prevProps.play !== play) {
  //     if (play) {
  //       this.playPromise = videoPlayer.play()
  //     } else if (this.playPromise !== undefined) {
  //       // eslint-disable-next-line no-void
  //       void this.playPromise.then(() => {
  //         videoPlayer.pause()
  //       })
  //     } else {
  //       videoPlayer.pause()
  //     }
  //   }
  //
  //   // Player position
  //   if (videoPlayer.currentTime !== position) {
  //     videoPlayer.currentTime = position
  //   }
  // }, [])

  return (
    <video
      crossOrigin="anonymous"
      ref={videoPlayer}
      preload="metadata"
      muted={false}
      className="c-video__element"
      poster={poster}
      controls={showControls}
    >
      <source src={src} type={type} />
      {children}
    </video>
  )
}

export default Vidiho
