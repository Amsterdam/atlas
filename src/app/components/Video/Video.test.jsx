import React from 'react'
import { shallow } from 'enzyme'
import Video from './Video'

describe('Video', () => {
  let component
  let componentInstance
  let videoPlayerMock

  beforeEach(() => {
    videoPlayerMock = {
      play: jest.fn(() => Promise.resolve()),
      pause: jest.fn(),
      currentTime: 0,
    }

    component = shallow(
      <Video poster="/assets/video/map.jpg" src="/assets/video/map.mp4" type="video/mp4" />,
    )
    componentInstance = component.instance()
    componentInstance.videoPlayer = videoPlayerMock
  })

  it('should render', () => {
    expect(component).toMatchSnapshot()
  })

  describe('componentDidUpdate', () => {
    it('should play the video', () => {
      component.setProps({
        play: true,
        position: 0,
      })
      expect(videoPlayerMock.play).toHaveBeenCalled()
      expect(videoPlayerMock.pause).not.toHaveBeenCalled()
    })

    it("should update the video position when play state doesn't change", () => {
      component.setProps({
        position: 10,
      })
      expect(videoPlayerMock.currentTime).toEqual(10)
    })

    it('should pause the video', async () => {
      component.setProps({
        play: true,
      })
      expect(videoPlayerMock.play).toHaveBeenCalled()
      expect(videoPlayerMock.pause).not.toHaveBeenCalled()
      expect(componentInstance.playPromise).toBeTruthy()
      videoPlayerMock.play.mockReset()
      videoPlayerMock.pause.mockReset()
      await component.setProps({
        play: false,
      })

      expect(videoPlayerMock.play).not.toHaveBeenCalled()
      expect(videoPlayerMock.pause).toHaveBeenCalled()
    })

    it('should pause the video in IE11', async () => {
      videoPlayerMock.play.mockReset()
      videoPlayerMock.pause.mockReset()
      component.setProps({
        play: true,
      })
      expect(videoPlayerMock.play).toHaveBeenCalled()
      expect(videoPlayerMock.pause).not.toHaveBeenCalled()
      componentInstance.playPromise = undefined
      videoPlayerMock.play.mockReset()
      videoPlayerMock.pause.mockReset()
      await component.setProps({
        play: false,
      })

      expect(videoPlayerMock.play).not.toHaveBeenCalled()
      expect(videoPlayerMock.pause).toHaveBeenCalled()
    })

    it('should do nothing when stopping the video while play is not defined', () => {
      videoPlayerMock.play.mockReset()
      videoPlayerMock.pause.mockReset()
      component.setProps({
        play: true,
        position: 0,
      })
      videoPlayerMock.play.mockReset()
      videoPlayerMock.pause.mockReset()

      component.setProps({
        play: false,
        position: 0,
      })
      expect(videoPlayerMock.play).not.toHaveBeenCalled()
    })
  })
})
