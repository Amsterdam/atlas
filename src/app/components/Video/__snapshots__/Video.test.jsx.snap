// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video should render 1`] = `
<video
  className="c-video__element"
  data-object-fit={true}
  loop={true}
  muted={true}
  poster="/assets/video/map.jpg"
  preload="metadata"
>
  <source
    src="/assets/video/map.mp4"
    type="video/mp4"
  />
</video>
`;
